src/mlpack/methods/cf/normalization/item_mean_normalization.hpp

```diff
@@ -89,11 +89,26 @@ class ItemMeanNormalization
    */
   void Normalize(arma::sp_mat& cleanedData)
   {
-    itemMean = arma::vec(arma::mean(cleanedData, 1));
-
+    // Calculate itemMean.
+    itemMean = arma::vec(cleanedData.n_rows, arma::fill::zeros);
+    arma::Col<size_t> ratingNum(cleanedData.n_rows, arma::fill::zeros);
     arma::sp_mat::iterator it = cleanedData.begin();
     arma::sp_mat::iterator it_end = cleanedData.end();
     for (; it != it_end; it++)
+    {
+      itemMean(it.row()) += *it;
+      ratingNum(it.row()) += 1;
+    }
+    for (size_t i = 0; i < itemMean.n_elem; i++)
+    {
+      if (ratingNum(i) != 0)
+        itemMean(i) /= ratingNum(i);
+    }
+
+    // Normalize the data.
+    it = cleanedData.begin();
+    it_end = cleanedData.end();
+    for (; it != it_end; it++)
```

Comment:
Looks like there is a bug here. Look at the valgrind output
```
==19199== Invalid read of size 8
==19199==    at 0x3DEA7C1: row (SpMat_bones.hpp:385)
==19199==    by 0x3DEA7C1: arma::SpMat<double>::const_iterator::operator!=(arma::SpMat<double>::const_iterator const&) const (SpMat_iterators_meat.hpp:239)
==19199==    by 0x4135BE7: mlpack::cf::ItemMeanNormalization::Normalize(arma::SpMat<double>&) (item_mean_normalization.hpp:111)
==19199==    by 0x41AA940: mlpack::cf::CFType<mlpack::cf::NMFPolicy, mlpack::cf::ItemMeanNormalization>::Train(arma::SpMat<double> const&, mlpack::cf::NMFPolicy const&, unsigned long, double, bool) (cf_impl.hpp:132)
==19199==    by 0x4194105: mlpack::cf::CFType<mlpack::cf::NMFPolicy, mlpack::cf::ItemMeanNormalization>::CFType<arma::SpMat<double> >(arma::SpMat<double> const&, mlpack::cf::NMFPolicy const&, unsigned long, unsigned long, unsigned long, double, bool) (cf_impl.hpp:72)
==19199==    by 0x41654F2: void CFTest::Serialization<mlpack::cf::NMFPolicy, mlpack::cf::ItemMeanNormalization>() (cf_test.cpp:441)
```
Looks like `it_end` becomes out of date as soon as you change the matrix. The following code should fix the issue
```
for (; it != cleanedData.end(); it++)
```

Reply:
Ahh thanks a lot!! I spent some time but was not able to debug this. BTW, where can I read the valgrind output?

Reply:
I ran valgrind on my local computer:) It isn't integrated with CI.